network/lwip2transport/tcp.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lwip2transport
+
+import (
+	"context"
+	"io"
+	"net"
+
+	"github.com/Jigsaw-Code/outline-internal-sdk/transport"
+	lwipLib "github.com/eycorsican/go-tun2socks/core"
+)
+
+type tcpHandler struct {
+	dialer transport.StreamDialer
+}
+
+// newTCPHandler returns a Shadowsocks lwIP connection handler.
+func newTCPHandler(client transport.StreamDialer) *tcpHandler {
+	return &tcpHandler{client}
+}
+
+func (h *tcpHandler) Handle(conn net.Conn, target *net.TCPAddr) error {
+	proxyConn, err := h.dialer.Dial(context.Background(), target.String())
+	if err != nil {
+		return err
+	}
+	// TODO: Request upstream to make `conn` a `core.TCPConn` so we can avoid this type assertion.
+	go relay(conn.(lwipLib.TCPConn), proxyConn)
+	return nil
+}
+
+func copyOneWay(leftConn, rightConn transport.StreamConn) (int64, error) {
+	n, err := io.Copy(leftConn, rightConn)
```

Comment:
Might be worth mentioning that this leverages the ReaderFrom/WriterTo is available.

Reply:
I feel this should be a common sense to go developers because [`io.Copy` already documented it](https://pkg.go.dev/io#Copy).

Reply:
That's an important performance aspect that's is worth calling out. Perhaps mention in the copyOneWay comment.

Reply:
added comment